nomt/src/store/writeout.rs

```diff
@@ -212,11 +307,83 @@ impl IoDmux {
 }
 
 struct Cx {
+    wal_write_out: WalWriteOut,
     bbn_write_out: BbnWriteOut,
     ln_write_out: LnWriteOut,
+    ht_write_out: HtWriteOut,
     meta_swap: Option<MetaSwap>,
 }
 
+struct WalWriteOut {
+    wal_fd: RawFd,
+    wal_blob: (*mut u8, usize),
+    wal_blob_pns: Vec<u32>,
+    remaining: usize,
+}
+
+impl WalWriteOut {
+    fn new(wal_fd: RawFd, wal_blob: (*mut u8, usize)) -> Self {
+        // HACK: we cannot issue one big write, so we split it into multiple pages.
+        let wal_blob_len = wal_blob.1;
+        assert!(wal_blob_len % PAGE_SIZE == 0);
+        let mut wal_blob_parts = Vec::new();
+        for i in 0..wal_blob_len / PAGE_SIZE {
+            wal_blob_parts.push(i as u32);
+        }
+        Self {
+            wal_fd,
+            wal_blob,
+            remaining: wal_blob_parts.len(),
+            wal_blob_pns: wal_blob_parts,
+        }
+    }
+
+    fn truncate_file(&mut self) {
+        unsafe {
+            let mut f = File::from_raw_fd(self.wal_fd);
+            f.set_len(0).unwrap();
+            f.seek(SeekFrom::Start(0)).unwrap();
+            let _ = f.into_raw_fd();
+        }
+    }
+
+    fn send_writes(&mut self, io: &mut IoDmux) {
+        // send all writes
+        while self.wal_blob_pns.len() > 0 {
+            // UNWRAP: we know the list is not empty.
+            let pn = self.wal_blob_pns.pop().unwrap();
+
+            unsafe {
+                let ptr = self.wal_blob.0.add(pn as usize * PAGE_SIZE);
+                let len = PAGE_SIZE;
+                if let Err(_) = io.try_send_wal(IoKind::WriteRaw(self.wal_fd, pn as u64, ptr, len))
```

Comment:
I'm not sure this is OK, because there are no guarantees that io_uring executes requests in order.
There are two things I think need to be true for this to be correct:
  1. That `pwrite` doesn't care about the actual offset when an fd is opened with  `O_APPEND` and always appends the page.
  2. That the WAL pages are ordering independent.
  
  
Quoting from https://linux.die.net/man/2/pwrite

> POSIX requires that opening a file with the O_APPEND flag should have no affect on the location at which pwrite() writes data. However, on Linux, if a file is opened with O_APPEND, pwrite() appends data to the end of the file, regardless of the value of offset. 

It seems that (1) is possibly guaranteed on linux but that this behavior can't be relied on for other POSIXes.

I see the need for a hack and understand the issue this is solving, but it seems better to just avoid the `io` subsystem altogether and do the write in a background thread or something.

Reply:
It's true that the pages will land in no particular order. However, given that they all land, the file will have the same contents regardless of the order they actually got written. The requirement for a successful sync is that the manifest gets updated and that only happens iff WAL is successfuly fsynced.

Reply:
I want to be clear: that is only true if 

> 1. pwrite doesn't care about the actual offset when an fd is opened with O_APPEND and always appends the page

Reply:
Ah, my bad! I was sure I committed the changes that removed O_APPEND but apparently somehow missed those.